### PR TITLE
basic rest functionality and db init

### PR DESCRIPTION
**Pull request by kmdzinarishvili**

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
     ports:
       - "5432:5432"
     environment:
-      POSTGRES_DB: wsmt
-      POSTGRES_USER: wsmt
-      POSTGRES_PASSWORD: wsmt
\ No newline at end of file
+      POSTGRES_DB: reminderApp
+      POSTGRES_USER: keti
+      POSTGRES_PASSWORD: cKiJ3tX3@hA2
```
Comment (avozhakov) [inline docker-compose.yaml]: There also should be your api (spring application) as service

File: pom.xml [hunk 2]
```diff
@@ -73,12 +73,9 @@
                 <version>9.19.4</version>
 
                 <configuration>
-                    <url>jdbc:postgresql://localhost:5432/</url>
-                    <user>wsmt</user>
-                    <password>wsmt</password>
-                    <schemas>
-                        <schema>wsmt</schema>
-                    </schemas>
```
Comment (avozhakov) [inline pom.xml:81]: Why do you need it here?

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/controllers/ReminderController.java [hunk 3]
```diff
@@ -1,17 +1,46 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
 
 @RestController
+@RequestMapping(path = "reminders")
+
 public class ReminderController {
+    private final ReminderService reminderService;
+
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/controllers/ReminderController.java:15]: You don't need this constructor here. Spring will inject all the dependencies itself. It's about controllers only

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/controllers/ReminderController.java [hunk 4]
```diff
@@ -1,17 +1,46 @@
...
-    public String getAll(){
-        return "all";
+        this.reminderService = reminderService;
     }
 
+    @GetMapping
+    public List<Reminder> listReminders() {
+        return reminderService.getReminders();
+    }
+
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/controllers/ReminderController.java:27]: I believe it should be like "/{id}"

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java [hunk 5]
```diff
@@ -1,5 +1,94 @@
...
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.Arrays;
+
+
 public class Reminder {
-    
+    private Long id;
+    private String title;
+    private RecurrenceType recurrence;
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java:19]: Use LocalDateTime

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java [hunk 6]
```diff
@@ -1,5 +1,94 @@
...
+
+    public Reminder(Long id, String title, RecurrenceType recurrence, LocalDate date, LocalTime time, byte[] attachment) {
+        this.id = id;
+        this.title = title;
+        this.recurrence = recurrence;
+        this.date = date;
+        this.time = time;
+        this.attachment = attachment;
+    }
+
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java]: You can use Lombok if you want

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java [hunk 7]
```diff
@@ -1,4 +1,58 @@
...
+    private final ReminderDAO reminderDao;
+
+    public ReminderService(ReminderDAO reminderDao) {
+        this.reminderDao = reminderDao;
+    }
+
+    public List<Reminder> getReminders() {
+        return reminderDao.selectReminders();
+    }
+
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java:25]: save and update methods should return saved ore updated dto with id from db

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -1,3 +1,3 @@
-spring.datasource.url=jdbc:postgresql://localhost:5432/wsmt
-spring.datasource.username=wsmt
-spring.datasource.password=wsmt
```
Comment (avozhakov) [inline src/main/resources/application.properties:3]: You can move these properties to .yaml and remove this .properties file

File: src/main/resources/db/migration/V1__initTables.sql [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+CREATE TABLE labels_reminders(
+    reminder_id INTEGER,
+    label_id INTEGER,
+    PRIMARY KEY(reminder_id, label_id)
+);
+
+CREATE TABLE users(
+    user_id serial PRIMARY KEY,
+    username VARCHAR ( 50 ) NOT NULL,
+    password varchar (255) NOT NULL,
```
Comment (avozhakov) [inline src/main/resources/db/migration/V1__initTables.sql:26]: Add email field

File: src/main/resources/db/migration/V1__initTables.sql [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    reminder_id INTEGER,
+    user_id INTEGER,
+    PRIMARY KEY(reminder_id, user_id)
+);
+
+
+CREATE TABLE modifications (
+    reminder_id serial PRIMARY KEY,
+	title VARCHAR ( 255 ) NOT NULL,
+	recurrence VARCHAR ( 50 ) NOT NULL,
```
Comment (avozhakov) [inline src/main/resources/db/migration/V1__initTables.sql:41]: DateTime as well

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java (hunk 5 above)
Comment (kmdzinarishvili) [reply src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java:19]: I was looking at the requirements and the user should have a time zone. Would that require a different datatype like timestamp? Or is it just for the user to have it linked with their profile, and changing it doesn't impact the time of the reminders?

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java (hunk 7 above)
Comment (kmdzinarishvili) [reply src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java:25]: I had a question about dtos. I usually use them when getting data from the user (with the @RequestBody annotation). Should I be using them to send and receive data? Should the dto also have an id attribute? Or were you referring to the fact that I should define something like an ApiResponse class and return that?

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java (hunk 5 above)
Comment (avozhakov) [reply src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java:19]: I'm thinking about the time zones. Will write the approach soon

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java (hunk 7 above)
Comment (avozhakov) [reply src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java:25]: The best practice here is:
Use DTOs (requests and responses) for the controller layer
Use models for the service layer
User entity for the repository layer
And map one to another using mapper.

Such as you are not using ORM and repository you can do something like:
Use the same DTOs (Reminder) in the service and DAO layer 
Use (ReminderRequest / ReminderResponse) DTOs in the controller layer

It means: 
Create ReminderRequest and use it as a request parameter for the controller endpoints.
pass it to the service layer and then map it to the Reminder model
after some work service layer returns that or a new Reminder from service
map that Reminder in the controller to ReminderResponse and return from the endpoint

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java [hunk 11]
```diff
@@ -1,4 +1,47 @@
...
+    public Reminder addNewReminder(Reminder reminder) {
+        return reminderDao.insertReminder(reminder);
+    }
+
+    public Reminder updateReminder(Reminder reminder) {
+        if(!reminderDao.isIdValid(reminder.getId())){
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java]: You should just update reminder here inside the try-catch block and if there is any exception then throw IllegalArgumentException with the caught exception message 
Your dao layer should throw database exceptions

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java [hunk 12]
```diff
@@ -1,4 +1,47 @@
...
+            throw new InvalidInputException("No reminder with this id");
+        }
+       return reminderDao.updateReminder(reminder);
+
+    }
+
+    public Boolean deleteReminder(Long id) {
+        if(!reminderDao.isIdValid(id)){
```
Comment (avozhakov) [inline src/main/java/com/lineate/mdzinarishvili/reminderapp/services/ReminderService.java]: The same here as above

File: src/main/resources/db/migration/V1.1__separateRecurrence.sql [hunk 13]
```diff
@@ -0,0 +1,15 @@
+create table recurrence_types (
```
Comment (avozhakov) [inline src/main/resources/db/migration/V1.1__separateRecurrence.sql]: Just use V1_, V2_, V3_.. prefixes

File: src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java (hunk 5 above)
Comment (avozhakov) [reply src/main/java/com/lineate/mdzinarishvili/reminderapp/models/Reminder.java:19]: Use DateTime in java, And UTC timestamp in the db

**Review by avozhakov — APPROVED** [review]

**Pull request merged**